activerecord/lib/active_record/insert_all.rb

```diff
@@ -0,0 +1,185 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  class InsertAll
+    def initialize(model, inserts, **options)
+      raise ArgumentError, "inserts should be an array" unless inserts.is_a?(Array)
+      @model, @connection, @options = model, model.connection, options
+      @keys, @values_list = connection.prepare_keys_and_values_for_insert(inserts, model.table_name)
+    end
+
+    def execute
+      if values_list.empty?
+        ActiveRecord::Result.new([], [])
+      else
+        connection.exec_query to_sql, "Bulk Insert"
+      end
+    end
+
+    def to_sql
+      sql_builder.build_sql(
+        model,
+        keys,
+        values_list,
+        returning: returning,
+        on_duplicate: on_duplicate,
+        unique_by: unique_by,
+        updatable_columns: updatable_columns)
+    end
+
+    private
+      attr_reader :model, :connection, :options, :keys, :values_list
+
+      def sql_builder
+        connection.insert_all_sql_builder
+      end
+
+      def returning
+        returning = options.fetch(:returning) { connection.supports_insert_returning? && primary_keys }
+        returning = false if returning == []
+
+        if returning && !connection.supports_insert_returning?
+          raise ArgumentError, "#{connection.class} does not support :returning"
+        end
```

Comment:
Replace with `ensure_connection_supports_returning` and turn the whole thing into a single conditional.

Reply:
(This comment was on the lines above when they were part of the `returning` method: when each option — instead of being an instance variable — was a method that fetched its value from `options` _and_ did it's own validation.)

Do you think it'd be good to split this method further — like this?

```ruby
def ensure_valid_options_for_connection!
  ensure_valid_value_for_on_duplicate!
  ensure_connection_supports_returning! if returning
  ensure_connection_supports_skipping_duplicates! if on_duplicate == :skip
  ensure_connection_supports_updating_duplicates! if on_duplicate == :update
  ensure_connection_supports_conflict_target! if unique_by
end
```

Reply:
I'd say it's fine how it is now.